sound/soc/sof/core.c

```diff
@@ -53,29 +54,76 @@ static const struct sof_panic_msg panic_msg[] = {
 };
 
 #if IS_ENABLED(CONFIG_SND_SOC_SOF_CLIENT)
+static int snd_sof_register_ipcflood_test(struct snd_sof_dev *sdev)
+{
+	int i;
+	int ret;
+
+	if (!IS_ENABLED(CONFIG_SND_SOC_SOF_DEBUG_IPC_FLOOD_TEST))
+		return 0;
+
+	for (i = 0; i < CONFIG_SND_SOC_SOF_DEBUG_IPC_FLOOD_TEST_NUM; i++) {
+		ret = sof_client_dev_register(sdev, "ipc_test", i);
+		if (ret < 0)
+			break;
+	}
+
+	if (ret) {
+		for (; i >= 0; --i)
+			sof_client_dev_unregister(sdev, "ipc_test", i);
+	}
+
+	return ret;
+}
+
+static void snd_sof_unregister_ipcflood_test(struct snd_sof_dev *sdev)
+{
+	int i;
+
+	if (!IS_ENABLED(CONFIG_SND_SOC_SOF_DEBUG_IPC_FLOOD_TEST))
+		return;
+
+	for (i = 0; i < CONFIG_SND_SOC_SOF_DEBUG_IPC_FLOOD_TEST_NUM; i++)
+		sof_client_dev_unregister(sdev, "ipc_test", i);
+}
+
 static int snd_sof_register_clients(struct snd_sof_dev *sdev)
 {
+	int ret;
+
+	/* Register platform independent client devices */
+	ret = snd_sof_register_ipcflood_test(sdev);
+	if (ret) {
+		dev_err(sdev->dev, "error: IPC flood test client registration failed\n");
+		return ret;
+	}
+
+	/* Platform depndent client device registration */
 	if (sof_ops(sdev) && sof_ops(sdev)->register_clients)
-		return sof_ops(sdev)->register_clients(sdev);
+		ret = sof_ops(sdev)->register_clients(sdev);
 
-	return 0;
+	if (ret)
+		snd_sof_unregister_ipcflood_test(sdev);
+
+	return ret;
 }
 
 static void snd_sof_unregister_clients(struct snd_sof_dev *sdev)
 {
 	if (sof_ops(sdev) && sof_ops(sdev)->unregister_clients)
 		sof_ops(sdev)->unregister_clients(sdev);
+
+	snd_sof_unregister_ipcflood_test(sdev);
```

Comment:
why is this all in core.c? This doesn't seem right to me.

Reply:
It is a generic client and does not belong to platform code, it has to be in the core.
We had it in the debug.c

We have introduced a callback for client registration and I think it is better to keep all client device creation, destruction in one place to minimize the chances of random ordering issues.

We could move the generic debug feature registrations to debug.c, if that makes more sense?

Reply:
...or to sof-client.c, maybe even move `snd_sof_register_clients()` and `snd_sof_unregister_clients()` there. Maybe even move all generic IPC-flood (de)registration to sof-client-ipc-test.c

Reply:
right, let me see how that will look like, thanks.